src/freertos_drivers/common/SN74HC595GPO.hxx

```diff
@@ -0,0 +1,242 @@
+/// std::atomic and requires an implementation of std::atomic on any platform
+/// that makes use of it. For example, at the time of creating this device
+/// driver, ARMv7-M (Cortex-M3/M4/M4F) does have support for std::atomic,
+/// while ARMv6-M (Cortex-M0) does not.
+/// @tparam N number of chips on the bus, starting at 1
+template <uint8_t N = 1> class SN74HC595 : public OSThread
+{
+public:
+    /// Constructor.
+    /// @param request_refresh_operation This is an application specific
+    ///        callback for triggering a sequence of events that will result
+    ///        in @ref refresh() being called. The exact implementation is
+    ///        application specific, and mostly is determined by whether or not
+    ///        these outputs must be toggled from an interrupt.
+    SN74HC595(void (*request_refresh_operation)(void))
+        : requestRefreshOperation_(request_refresh_operation)
+        , spiFd_(-1)
+        , sem_()
+        , ioPending_(false)
+    {
+        HASSERT(N > 0);
+        for (unsigned i = 0; i < N; ++i)
+        {
+            gpoData_[i] = 0;
+        }
+    }
+
+    /// Initialize the SN74HC595 settings. Typically called in hw_postinit(),
+    /// not hw_preinit() or hw_init().
+    /// @param spi_name spi interface that the SN74HC595 is on
+    void init(const char *spi_name)
+    {
+        spiFd_ = ::open(spi_name, O_WRONLY);
+        HASSERT(spiFd_ >= 0);
+
+        // configure SPI bus settings
+        uint8_t spi_mode = SPI_MODE_0;
+        uint8_t spi_bpw = 8;
+        uint32_t spi_max_speed_hz = SPI_MAX_SPEED_HZ;
+        ::ioctl(spiFd_, SPI_IOC_WR_MODE, &spi_mode);
+        ::ioctl(spiFd_, SPI_IOC_WR_BITS_PER_WORD, &spi_bpw);
+        ::ioctl(spiFd_, SPI_IOC_WR_MAX_SPEED_HZ, &spi_max_speed_hz);
+
+        requestRefreshOperation_();
+
+        // start the thread at the highest priority in the system
+        start("SN74HC595", get_priority_max(), 512);
+    }
+
+    /// Triggers the helper thread to wakeup and refresh the outputs.
+    /// @param from_isr true if called from an ISR.
+    void refresh(bool from_isr = false)
+    {
+        if (!ioPending_)
+        {
+            ioPending_ = true;
+            if (from_isr)
+            {
+                int woken;
+                sem_.post_from_isr(&woken);
+            }
+            else
+            {
+                sem_.post();
+            }
+        }
+    }
+
+private:
+    /// maximum SPI clock speed in Hz
+    static constexpr uint32_t SPI_MAX_SPEED_HZ = 1000000;
+
+    /// User entry point for the created thread.
+    /// @return exit status
+    void *entry() override
+    {
+        for ( ; /* forever */ ; )
+        {
+            sem_.wait();
+            if (ioPending_)
+            {
+                ioPending_ = false;
+                for (unsigned i = 0; i < N; ++i)
+                {
+                    uint8_t data = gpoData_[0];
+                    ::write(spiFd_, &data, 1);
```

Comment:
why aren't we writing all bytes in one go?
Specifically, this will not work for more than 1 chip, because the latch gets toggled inbetween, the mutex gets released and other SPI transaction can happen, sending garbage to the remaining shift registers.

Reply:
Fixed.